geomesa-gs-monitor-elasticsearch/src/main/scala/org/geomesa/gs/monitor/elastic/ElasticRequestDataListener.scala

```diff
@@ -8,137 +8,165 @@
-    }
-    catch{
-      case e: MalformedURLException => logger.error("Invalid URL " + u + ". Could not convert from string to URL. Trying any additional URLs")
+
+  import org.geomesa.gs.monitor.elastic.ElasticRequestDataListener._
+
+  private val config = ConfigFactory.load.getConfig(GEOSERVER_MONITOR_ELASTICSEARCH_KEY)
+  private val index = config.getString("index")
+
+  private val excludedFields =
+    Option("excludedFields")
+      .filter(config.hasPath)
+      .map(config.getStringList)
+      .map(_.asScala.toSet)
+      .getOrElse(Set.empty)
+  private val timeout =
+    Option("timeout")
+      .filter(config.hasPath)
+      .map(config.getLong)
+      .filter(ms => ms > 1000L && ms < 1000000L)
+      .getOrElse(10000L)
+
+  private val gson = ExtendedRequestData.getGson(excludedFields)
+  private val restClient = getRestClient(config)
+
+  /*
+   * GeoServer Request Monitor Timeout Strategy
+   *
+   * If a request times out, it never reaches the completed or post-processed callbacks and thus
+   * would never be sent to Elasticsearch. To combat this, all in-progress requests are stored
+   * in a set ordered by their start time. Whenever a request is post-processed, it is sent to
+   * Elasticsearch and removed from the set. Any in-progress requests that exceed the configured
+   * timeout are marked as failed and sent to Elasticsearch. If a request completes after it has
+   * already been marked as a timeout, the Elasticsearch index is updated using the request id.
+   */
+
+  private val writeQueue = new LinkedBlockingQueue[RequestData]
+  private val requestsInProgress = new SortedHashSet(RequestData.startTimeOrdering)
+
+  private val elasticsearchWriter = new Runnable {
+    override def run(): Unit = {
+      while (!Thread.interrupted) {
+        val rd = writeQueue.take // block until data is available to send
+
+        try {
+          putElasticsearch(new ExtendedRequestData(rd))
+          logger.debug(s"Sent request '${rd.uid}' to Elasticsearch")
+        } catch {
+          case ex: Exception => logger.error(s"Failed to send request '${rd.uid}' to Elasticsearch: ${ex.getMessage}")
+        }
+      }
     }
   }
-  if (hostList.length == 0) {
-    logger.error("No URL given. Could not resolve " + envVars.mkString(","))
-    throw new Exception("Given URL(s) " + envVars.mkString(",") + " cannot be read by Java URL")
-  }
-
-  val hosts = hostList.toArray
-  val client = new RestHighLevelClient(
-    RestClient.builder(hosts:_*)
-  )
-  logger.debug("Sending requests to " + hosts)
 
-  var index = sys.env.getOrElse("GEOSERVER_ES_INDEX", null)
-  if (index == null){
-    index = "geoserver"
-    logger.warn("No index name provided. Index will be set to default name 'geoserver'")
+  private val timeoutHandler = new Runnable {
+    override def run(): Unit = {
+      // don't read in-progress requests while they are being updated to maintain ordering
+      requestsInProgress.synchronized {
```

Comment:
are you worried about contention here at all? i can think of some non-blocking ways to implement this but they might be kind of complicated.

Reply:
i wonder if it might be better to just use a ConcurrentHashSet, and iterate over all the entries instead of trying to sort them by timeout. that wouldn't require any blocking, although you can still have contention on concurrent collections.

Reply:
there are some methods in ConcurrentHashSet that you can use to avoid that. something like:
```
  val hs = new ConcurrentHashMap[String, Object]()
  val expired = hs.asScala.filter(_ => true /*check for expiration*/)
  expired.foreach { case (k, v) =>
    if (hs.remove(k, v)) {
      /* this is an expired entry that hasn't been updated since we checked it */
    }
  }
```

Reply:
when the request completes, you'd have to do a similar `remove` check - only one of those will succeed, so you can handle it appropriately from there based on whether it expired or completed first.

Reply:
You can also create a "CocurrentHashSet", but you have to call `java.util.Collections.newSetFromMap(new ConcurrentHashMap[Object, Boolean])`

Reply:
So in each of the listener callbacks I need to update the RequestData object in the set because it might have changed, but I also need to implement equals such that I can remove the correct element. Is the best way to work around this using a Map[RequestData, RequestData]?

Reply:
could you make it a `Map[String, RequestData]`, keyed on the request id?

Reply:
Also I think there still might be an issue with the logic. For the timed out requests, we only want to add it to the queue if it hasn't already ready been removed. But for the completed requests we want to add it either way, but it needs to be after the timed out request, and I don't know how to guarantee that it will be (synchronize around adds to the queue?)

Reply:
I think you could use `ConcurrentHashMap.compute` for that. that method is atomic (and may block), and takes a key, and a function that operates on the key, the old value, and returns a new value or null to remove the item. so for the completed requests, you call compute and in the function insert the value into the write queue and return null to remove the item. for the timed out requests, you do the same thing but you check to make sure that the old value is not null, which indicates it was already removed by the completed request. that would guarantee ordering, if i'm thinking about it right